File: fsharp/focl/p061.fs
module Focl.P061

open Xunit
open FsUnit.Xunit

(*
https://4clojure.oxal.org/#/problem/61
Problem 61, Map Construction
Difficulty: easy
Write a function which takes a vector of keys 
and a vector of values and constructs a map from them.
    (= (__ [:a :b :c] [1 2 3]) {:a 1, :b 2, :c 3})
    (= (__ [1 2 3 4] ["one" "two" "three"]) {1 "one", 2 "two", 3 "three"})
    (= (__ [:foo :bar] ["foo" "bar" "baz"]) {:foo "foo", :bar "bar"})
*)

let la2, lb2, r2 =
    [ 1; 2; 3; 4 ], [ "one"; "two"; "three" ], [ (1, "one"); (2, "two"); (3, "three") ]

let f061a la lb =
    let n = ([ List.length la ] @ [ List.length lb ] |> List.min) - 1
    List.zip la[..n] lb[..n]

[<Fact>]
let t061a () = f061a la2 lb2 |> should equal r2

let f061b (a: 'a list) (b: 'b list) : Map<'a, 'b> =
    let rec f (acc: Map<'a, 'b>) (a: 'a list) (b: 'b list) =
        match a, b with
        | [], _ -> acc
        | _, [] -> acc
        | x :: xs, y :: ys -> f (acc.Add(x, y)) xs ys

    f Map.empty a b

[<Fact>]
let t061b () =
    let exp = Map.ofList [ (1, "one"); (2, "two"); (3, "three") ]
    f061b [ 1; 2; 3; 4 ] [ "one"; "two"; "three" ] |> should equal exp

File: clojure/src/ace/focl/p061.clj
(ns ace.focl.p061
  (:require
   [clojure.test :as t]))

;; Write a function which takes a vector of keys and
;; a vector of values and constructs a map from them.

(defn run [f]
  (and (= (f [:a :b :c] [1 2 3])
          {:a 1, :b 2, :c 3})
       (= (f [1 2 3 4] ["one" "two" "three"])
          {1 "one", 2 "two", 3 "three"})
       (= (f [:foo :bar] ["foo" "bar" "baz"])
          {:foo "foo", :bar "bar"})))

(defn a [coll1 coll2]
  (let [shorter (min (count coll1) (count coll2))]
    (loop [i 0
           m {}]
      (if (<= shorter i)
        m
        (recur (inc i)
               (conj m {(get coll1 i) (get coll2 i)}))))))

(defn b [coll1 coll2]
  (apply hash-map (interleave coll1 coll2)))

(defn c [a b]
  (let [r (range (min (count a) (count b)))]
    (apply hash-map (mapcat (fn [n] [(get a n) (get b n)]) r))))

(defn d [a b]
  (into {} (map (fn [x y] {x y}) a b)))

(defn e [a b]
  (if (or (empty? a) (empty? b))
    {}
    (conj {(first a) (first b)}
          (e (rest a) (rest b)))))

(defn f [a b]
  (apply merge (map #(into {%1 %2}) a b)))

(t/deftest all
  (t/are [f] (run f)
    a b c d e f))
